### PR TITLE
Handle signals properly when dropping GVL

### PR DESCRIPTION
**Pull request by murgatroid99**
This is an improvement on #6472. We now no longer have extra ruby code to handle signals, we no longer spawn an extra thread to handle signals, and it no longer takes up to a second to handle a signal. Instead, we now use a built in facility to temporarily stop plucking when a signal arrives.

### CODE REVIEW

**Comment by murgatroid99** [comment]
All of the Ruby tests pass.

File: src/ruby/ext/grpc/rb_completion_queue.c [hunk 1]
```diff
@@ -158,8 +189,23 @@ grpc_event grpc_rb_completion_queue_pluck_event(VALUE self, VALUE tag,
...
   next_call.event.type = GRPC_QUEUE_TIMEOUT;
-  rb_thread_call_without_gvl(grpc_rb_completion_queue_pluck_no_gil,
-                             (void *)&next_call, NULL, NULL);
+  /* Loop until we finish a pluck without an interruption. The internal
+     pluck function runs either until it is interrupted or it gets an
+     event, or time runs out.
+
+     The basic reason we need this relatively complicated construction is that
+     we need to re-acquire the GVL when an interrupt comes in, so that the ruby
+     interpeter can do what it needs to do with the interrupt. But we also need
```
Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_completion_queue.c]: typo: interpeter

unfinished sentence

File: src/ruby/ext/grpc/rb_completion_queue.c [hunk 2]
```diff
@@ -52,21 +52,47 @@ typedef struct next_call_stack {
...
   void *tag;
+  volatile int interrupted;
 } next_call_stack;
 
 /* Calls grpc_completion_queue_next without holding the ruby GIL */
 static void *grpc_rb_completion_queue_next_no_gil(void *param) {
   next_call_stack *const next_call = (next_call_stack*)param;
-  next_call->event =
-      grpc_completion_queue_next(next_call->cq, next_call->timeout, NULL);
+  gpr_timespec increment = gpr_time_from_millis(20, GPR_TIMESPAN);
```
Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_completion_queue.c]: do we really have to wake up 50x a second?

File: src/ruby/ext/grpc/rb_completion_queue.c [hunk 3]
```diff
@@ -52,21 +52,47 @@ typedef struct next_call_stack {
...
+  gpr_timespec deadline;
+  do {
+    deadline = gpr_time_add(gpr_now(GPR_CLOCK_REALTIME), increment);
```
Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_completion_queue.c]: in next_call->timeout is in 19msec and our deadline is in 20, we just skip?

File: src/ruby/ext/grpc/rb_completion_queue.c (hunk 2 above)
Comment (murgatroid99) [inline src/ruby/ext/grpc/rb_completion_queue.c]: It's a time period that's short from the point of view of a human, but long in terms of processing time. In the future, we can tune it if it turns out to have a significant performance impact.

File: src/ruby/ext/grpc/rb_completion_queue.c (hunk 1 above)
Comment (murgatroid99) [inline src/ruby/ext/grpc/rb_completion_queue.c]: Fixed

File: src/ruby/ext/grpc/rb_completion_queue.c (hunk 3 above)
Comment (murgatroid99) [inline src/ruby/ext/grpc/rb_completion_queue.c]: Fixed. I moved the check to after `pluck` finishes.

**Comment by murgatroid99** [comment]
In addition to those fixes, I also changed my mind again and removed the ruby-level signal handling. The code I have seems to do fine at interrupting `pluck` with an `Interrupt` exception when I press Ctrl+C.

File: src/ruby/ext/grpc/rb_completion_queue.c (hunk 2 above)
Comment (jtattermusch) [inline src/ruby/ext/grpc/rb_completion_queue.c]: I see the point, but in general I think these kinds of tricks are problematic. It's probably fine with me but I really wish there was a cleaner solution.

**Comment by jtattermusch** [comment]
LGTM.

**Comment by murgatroid99** [comment]
OK, I think this can be merged. There was a problem that I was looking into, but it appears to not be related to this change.

File: src/ruby/ext/grpc/rb_completion_queue.c (hunk 2 above)
Comment (murgatroid99) [inline src/ruby/ext/grpc/rb_completion_queue.c]: Optimally, there would be a function to wake/fail an existing call to `pluck`. But we don't have that, and this is the best option I can see.

**Pull request merged**